### PR TITLE
Minor: Improve `HashJoinExec` documentation

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Related to https://github.com/apache/arrow-datafusion/issues/7949

## Rationale for this change

To explain https://github.com/apache/arrow-datafusion/issues/7949 I needed to explain the background of how HashJoinExec works, and so I wanted to encode that information into the documentation

## What changes are included in this PR?

Improved documentation. For example:

Screenshot 2023-10-27 at 11 45 38 AM

## Are these changes tested?

N/A

## Are there any user-facing changes?
Better `HashJoinExec` documentation.

### CODE REVIEW

**Review by andygrove — APPROVED** [review]
:heart:

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 1]
```diff
@@ -78,29 +77,140 @@ use futures::{ready, Stream, StreamExt, TryStreamExt};
...
+///           ...                                                       ...
+///     ┌────────────┐                                            ┌────────────┐
+///     │RecordBatch │                                            │RecordBatch │
+///     └────────────┘                                            └────────────┘
+///
+///        build side                                                probe side
+///
+/// ```
+///
+/// # Exampe "Optimial" Plans
```
Comment (viirya) [inline datafusion/physical-plan/src/joins/hash_join.rs:160]:
```suggestion
/// # Example "Optimal" Plans
```

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 2]
```diff
@@ -78,29 +77,140 @@ use futures::{ready, Stream, StreamExt, TryStreamExt};
...
+///
+/// The differences in the inputs means that for  classic "Star Schema Query",
```
Comment (viirya) [inline datafusion/physical-plan/src/joins/hash_join.rs:162]:
```suggestion
/// The differences in the inputs means that for classic "Star Schema Query",
```

File: datafusion/physical-plan/src/joins/hash_join.rs [hunk 3]
```diff
@@ -78,29 +77,140 @@ use futures::{ready, Stream, StreamExt, TryStreamExt};
...
+/// the optimal plan will be a **"Right Deep Tree"** . A Star Schema Query is
+/// one where there is one large table and several smaller "dimension" tables,
+/// joined on `Foreign Key = Primary Key` predicates. with predicates.
```
Comment (viirya) [inline datafusion/physical-plan/src/joins/hash_join.rs:165]:
```suggestion
/// joined on `Foreign Key = Primary Key` predicates.
```

**Review by viirya — APPROVED** [review]
Nice diagrams!

**Pull request merged**